### PR TITLE
New ppo

### PR DESCRIPTION
**Pull request by domenicrosati**
## Summary by Sourcery

Refactor the PPO-LLM agent by improving token padding, streamlining actor and critic networks, integrating entropy regularization, and refining trajectory management

Enhancements:
- Ensure the actor’s tokenizer assigns a pad_token when missing to prevent padding issues
- Default actor.forward to compute gradients, adjust state extraction, and normalize summed logits to avoid overflow
- Replace critic’s attention-based pooling with a GRU encoder over embeddings for state value estimation
- Introduce compute_actions helper to retrieve log probabilities and entropies and incorporate an entropy bonus into the PPO loss
- Revise episode buffering per environment with proper end-of-episode handling and lower the default max segment length to 8
- Log GPU memory allocation and reservation after each PPO update
- Change default number of parallel environments from 8 to 1 in the training configuration

### CODE REVIEW

File: src/agent/ppo_llm.py [hunk 1]
```diff
@@ -171,60 +161,90 @@ def create_conditional_logprobs(
         else:
             return conditional_logprobs
 
-
 class CriticNetwork(nn.Module):
```
[the remark this answers is not in the document]
Comment (ss26) [reply src/agent/ppo_llm.py:140]: probably a cleaner implementation of what @nmitrani suggested

File: src/agent/ppo_llm.py [hunk 2]
```diff
@@ -394,46 +404,63 @@ def update(self):
...
-            returns = torch.tensor(
-                [t.returns for t in episode], device=self.device, dtype=torch.float32
-            )
-
-            logprobs = torch.stack(
-                [
-                    self.actor.create_distribution(s, va, self.device).log_prob(a)
+            # 4 updates
+            num_updates = 4
+            for update in range(num_updates):
```
Comment (ss26) [inline src/agent/ppo_llm.py:409]: likely increasing sample efficiency and losses

File: src/env/test.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
+import random
+import json
+
+class SuperEasyEnv:
+    """
+    A simple environment where the agent must press a button, then say hello
```
Comment (ss26) [inline src/env/test.py:6]: lmao this env is cute af

File: src/agent/ppo_llm.py [hunk 4]
```diff
@@ -113,8 +104,7 @@ def forward(self, state: str, actions: list[str], device, requires_grad=False):
         else:
             with torch.no_grad():
                 outputs = self.model(**inputs)
-        logits = outputs.logits
-        logits = torch.clamp(logits, min=-100, max=100)
```
[the remark this answers is not in the document]
Comment (nmitrani) [reply src/agent/ppo_llm.py:117]: Gotta agree here, I think a default param in the forward method: (..., temperature=100.0): [...] action_token_logits.sum(dim=-1)/temperature
should work just fine. This is so we know what it is in the future and if someone else wants to contribute to this code, for them to find it quickly

File: src/agent/ppo_llm.py [hunk 5]
```diff
@@ -132,7 +122,7 @@ def forward(self, state: str, actions: list[str], device, requires_grad=False):
         ]
         # action_token_logprobs = action_token_logprobs_ * action_tokens["attention_mask"]
         action_token_logits = action_token_logits * action_tokens["attention_mask"]
-        return action_token_logits.sum(dim=-1)
+        return action_token_logits.sum(dim=-1) / 100.0
```
[the remark this answers is not in the document]
Comment (nmitrani) [reply src/agent/ppo_llm.py:125]: Gotta agree here, I think a default param in the forward method: (..., temperature=100.0): [...] action_token_logits.sum(dim=-1)/temperature
should work just fine. This is so we know what it is in the future and if someone else wants to contribute to this code, for them to find it quickly. Or a parameter in the __init__ with a default value of 100 could work as well if we don't want to change temperature between forward passes

**Review by nmitrani — APPROVED** [review]
